<a href="https://colab.research.google.com/github/ahmadfadhilnugraha/Bank_Loan/blob/main/Sprint_2_Bank_Loan.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assessing Default Risk in Loan Repayment

# Content <a id='back'></a>

* [Introduction](#Introduction)
* [Step 1. Data Review](#Step-1-Data-Review)
* [Step 2. Data Transformation](#Step-2-Data-Transformation)
    * [Step 2.1. Fixing Spelling Errors in the Education Column Name](#Step-2.1-Fixing-Spelling-Errors-in-the-Education-Column-Name)
    * [Step 2.2. Dropping Invalid Values in the Children Column](#Step-2.2-Dropping-Invalid-Values-in-the-Children-Column)
    * [Step 2.3. Addressing Anomalies in days_employed Column](#Step-2.3-Addressing-Anomalies-in-days_employed-Column)
    * [Step 2.4. Dropping Abnormal Values in dob_years Column](#Step-2.4-Dropping-Abnormal-Values-in-dob_years-Column)
    * [Step 2.5. Remove Abnormal Values in gender Column](#Step-2.5-Remove-Abnormal-Values-in-gender-Column)
    * [Step 2.6. Check for Duplicate Data](#Step-2.6-Check-for-Duplicate-Data)
    * [Step 2.7. Brief Summary of Data Transformation](#Step-2.7-Brief-Summary-of-Data-Transformation)
* [Step 3. Working with Missing Values](#Step-3-Working-with-Missing-Values)
    * [Step 3.1. Simplifying Data Processing with Family Status and Education IDs](#Step-3.1-Simplifying-Data-Processing-with-Family-Status-and-Education-IDs)
    * [Step 3.2. Fixing Missing Values in total_income](#Step-3.2-Fixing-Missing-Values-in-total_income)
    * [Step 3.3. Fixing Values in days_employed Column](#Step-3.3-Fixing-Values-in-days_employed-Column)
* [Step 4. Data Categorization](#Step-4-Data-Categorization)
    * [Step 4.1. Children Column Categorization](#Step-4.1-Children-Column-Categorization)
    * [Step 4.2. total_income Column Categorization](#Step-4.2-total_income-Column-Categorization)
* [Step 5. Hypotheses Testing](#Step-5-Hypotheses-Testing)
    * [Step 5.1. Hypothesis 1: Correlation between Having Children and Probability of Loan Default](#Step-5.1-Hypothesis-1-Correlation-between-Having-Children-and-Probability-of-Loan-Default)
    * [Step 5.2. Hypothesis 2: Correlation between Family Status and Timely Repayment](#Step-5.2-Hypothesis-2-Correlation-between-Family-Status-and-Timely-Repayment)
    * [Step 5.3. Hypothesis 3: Correlation between Income Level and Timely Repayment](#Step-5.3-Hypothesis-3-Correlation-between-Income-Level-and-Timely-Repayment)
    * [Step 5.4. Hypothesis 4: Impact of Loan Purpose on Default Rate](#Step-5.4-Hypothesis-4-Impact-of-Loan-Purpose-on-Default-Rate)
* [Step 6. Conclusion](#Step-6-Conclusion)



## Introduction

Your task is to prepare a report for the credit division of a bank. You will investigate the impact of a customer's marital status and the number of children they have on the probability of default in loan repayments. The bank already has some data regarding the creditworthiness of its customers.

This report will be considered when making credit assessments for potential customers. Credit assessments are crucial in evaluating the ability of prospective borrowers to repay their loans.

## Goals

In this meticulous data science project, I am delving into a comprehensive analysis of credit risk, specifically focusing on factors such as a customer's marital status and the number of children they have. These factors have the potential to significantly influence a customer's ability to settle loans and impact the probability of default in loan payments. Therefore, it is imperative to comprehend how these factors affect credit risk and take appropriate measures to mitigate such risks. As part of this exploration, I am formulating and testing four hypotheses:

1. Correlation between Having Children and Probability of Loan Default: Investigating whether having children is correlated with an increased likelihood of loan default.

2. Correlation between Family Status and Timely Repayment: Exploring whether a customer's family status is correlated with their ability to make timely loan repayments.

3. Correlation between Income Level and Timely Repayment: Analyzing whether a customer's income level correlates with their likelihood of repaying loans on time.

4. Impact of Loan Purpose on Default Rate: Assessing how the stated purpose of a loan influences the default rate.

This project holds significance not only for academic purposes but also for practical applications within the banking sector, contributing to a deeper understanding of credit risk and aiding in the development of effective risk mitigation strategies.

## Steps

The data related to users behavior is stored in the file /datasets/credit_scoring_eng.csv

This projects will consist of 4 steps:

1. Data Exploration
2. Data Cleaning
3. Hypotheses Testing
4. Conclusion and Recommendation

## Step 1. Data Review

Data Description:

- `children` - number of children in the family
- `days_employed` - customer's work experience in days
- `dob_years` - customer's age in years
- `education` - customer's education level
- `education_id` - identifier for the customer's education level
- `family_status` - identifier for the customer's marital status
- `family_status_id` - marital status identifier
- `gender` - customer's gender
- `income_type` - type of employment
- `debt` - whether the customer has ever defaulted on a loan
- `total_income` - monthly income
- `purpose` - purpose of obtaining the loan


In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
import pandas as pd

In [3]:
df = pd.read_csv('/content/drive/MyDrive/DS/SPRINT_2/credit_scoring_eng.csv')

In [4]:
df.head()

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [6]:
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

In [7]:
df.isnull().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [8]:
df.loc[:,('days_employed','total_income')].sort_values(by='days_employed')

days_employed  total_income
16335  -18388.949901     29788.629
4299   -17615.563266     19609.719
7329   -16593.472817     19951.655
17838  -16264.699501      8198.235
16825  -16119.687737     14644.430
...              ...           ...
21489            NaN           NaN
21495            NaN           NaN
21497            NaN           NaN
21502            NaN           NaN
21510            NaN           NaN

[21525 rows x 2 columns]

After applying a filtering process to the table, it was found that there are 2174 missing rows in the 'days_employed' and 'total_income' columns. To determine whether this amount of missing data is significant, it is necessary to calculate the percentage of missing data relative to the total dataset. If the percentage of missing data is too high, it can potentially impact the data analysis, and filling in the missing data needs to be done cautiously.

During the process of filling in the missing data, it is important to consider factors that may influence the value of the data, such as the type of employment or customer characteristics. Additionally, an examination should be conducted to determine whether the missing data is dependent on other indicators, such as age or income. This step can help identify customer characteristics and obtain more accurate information about credit risk. However, a more detailed analysis is required to evaluate the missing data and determine the appropriate strategy for addressing the issue.

In [9]:
df_null = df[pd.isnull(df['days_employed'])]
df_null.head()

children  days_employed  dob_years            education  education_id  \
12         0            NaN         65  secondary education             1   
26         0            NaN         41  secondary education             1   
29         0            NaN         63  secondary education             1   
41         0            NaN         50  secondary education             1   
55         0            NaN         54  secondary education             1   

        family_status  family_status_id gender    income_type  debt  \
12  civil partnership                 1      M        retiree     0   
26            married                 0      M  civil servant     0   
29          unmarried                 4      F        retiree     0   
41            married                 0      F  civil servant     0   
55  civil partnership                 1      F        retiree     1   

    total_income                   purpose  
12           NaN         to have a wedding  
26           NaN                 education  
29           NaN    building a real estate  
41           NaN  second-hand car purchase  
55           NaN         to have a wedding

In [10]:
df_null.groupby('income_type').count()

children  days_employed  dob_years  education  education_id  \
income_type                                                                  
business            508              0        508        508           508   
civil servant       147              0        147        147           147   
employee           1105              0       1105       1105          1105   
entrepreneur          1              0          1          1             1   
retiree             413              0        413        413           413   

               family_status  family_status_id  gender  debt  total_income  \
income_type                                                                  
business                 508               508     508   508             0   
civil servant            147               147     147   147             0   
employee                1105              1105    1105  1105             0   
entrepreneur               1                 1       1     1             0   
retiree                  413               413     413   413             0   

               purpose  
income_type             
business           508  
civil servant      147  
employee          1105  
entrepreneur         1  
retiree            413

In [11]:
df_null['income_type'].value_counts(normalize=True)*100

employee         50.827967
business         23.367065
retiree          18.997240
civil servant     6.761730
entrepreneur      0.045998
Name: income_type, dtype: float64

In [12]:
df_null['dob_years'].value_counts(normalize=True)*100

34    3.173873
40    3.035879
42    2.989880
31    2.989880
35    2.943882
36    2.897884
47    2.713891
41    2.713891
30    2.667893
28    2.621895
58    2.575897
57    2.575897
54    2.529899
56    2.483901
38    2.483901
52    2.437902
37    2.437902
33    2.345906
39    2.345906
50    2.345906
43    2.299908
45    2.299908
49    2.299908
51    2.299908
29    2.299908
46    2.207912
55    2.207912
48    2.115915
44    2.023919
53    2.023919
60    1.793928
62    1.747930
61    1.747930
32    1.701932
64    1.701932
23    1.655934
27    1.655934
26    1.609936
59    1.563937
63    1.333947
25    1.057958
24    0.965961
65    0.919963
66    0.919963
21    0.827967
22    0.781969
67    0.735971
0     0.459982
68    0.413983
71    0.229991
69    0.229991
20    0.229991
70    0.137994
72    0.091996
19    0.045998
73    0.045998
Name: dob_years, dtype: float64

In [13]:
df_null['education'].value_counts(normalize=True)*100

secondary education    64.765409
bachelor's degree      22.815087
SECONDARY EDUCATION     3.081877
Secondary Education     2.989880
some college            2.529899
Bachelor's Degree       1.149954
BACHELOR'S DEGREE       1.057958
primary education       0.873965
Some College            0.321987
SOME COLLEGE            0.321987
Primary Education       0.045998
PRIMARY EDUCATION       0.045998
Name: education, dtype: float64

In [14]:
df_null['family_status'].value_counts(normalize=True)*100

married              56.899724
civil partnership    20.331187
unmarried            13.247470
divorced              5.151794
widow / widower       4.369825
Name: family_status, dtype: float64

In [15]:
df_null['purpose'].value_counts(normalize=True)*100

having a wedding                            4.231831
to have a wedding                           3.725851
wedding ceremony                            3.495860
construction of own property                3.449862
housing transactions                        3.403864
buy real estate                             3.311868
purchase of the house for my family         3.265869
transactions with my real estate            3.265869
transactions with commercial real estate    3.219871
housing renovation                          3.219871
buy commercial real estate                  3.081877
buying property for renting out             2.989880
property                                    2.851886
real estate transactions                    2.805888
buy residential real estate                 2.805888
housing                                     2.759890
building a property                         2.713891
cars                                        2.621895
going to university                         2.

Data transformation is required by removing duplicate entries in the education column and calculating the average to address missing values in the dob_years column. Additionally, it is important to examine and rectify inconsistencies in the recording of the income_type and purpose columns. Subsequently, data source verification is necessary to ensure the accuracy of the data used in the analysis.

## Step 2. Data Transformation

### Step 2.1. Fixing Spelling Errors in the Education Column Name

In [16]:
df_edu_sorted = df.sort_values(by='education')
df_edu_sorted['education'].unique()

array(["BACHELOR'S DEGREE", "Bachelor's Degree", 'GRADUATE DEGREE',
       'Graduate Degree', 'PRIMARY EDUCATION', 'Primary Education',
       'SECONDARY EDUCATION', 'SOME COLLEGE', 'Secondary Education',
       'Some College', "bachelor's degree", 'graduate degree',
       'primary education', 'secondary education', 'some college'],
      dtype=object)

Several spelling errors, such as incorrect capitalization, were identified in the education column. This may lead to inaccuracies during data analysis.

In [17]:
def replace_wrong_education(wrong_edus, correct_edu):
    for wrong_edu in wrong_edus:
        df['education'] = df['education'].replace(wrong_edu,correct_edu)

duplicates_bach = ["BACHELOR'S DEGREE", "Bachelor's Degree"]
correct_bach = "bachelor's degree"
duplicates_grad = ['GRADUATE DEGREE', 'Graduate Degree']
correct_grad = 'graduate degree'
duplicates_prim = ['PRIMARY EDUCATION', 'Primary Education']
correct_prim = 'primary education'
duplicates_seco = ['SECONDARY EDUCATION', 'Secondary Education']
correct_seco = 'secondary education'
duplicates_some = ['SOME COLLEGE', 'Some College']
correct_some = 'some college'

replace_wrong_education(duplicates_bach, correct_bach)
replace_wrong_education(duplicates_grad, correct_grad)
replace_wrong_education(duplicates_prim, correct_prim)
replace_wrong_education(duplicates_seco, correct_seco)
replace_wrong_education(duplicates_some, correct_some)

In [18]:
df_edu_sorted = df.sort_values(by='education')
df_edu_sorted['education'].unique()

array(["bachelor's degree", 'graduate degree', 'primary education',
       'secondary education', 'some college'], dtype=object)

### Step 2.2. Dropping Invalid Values in the Children Column

In [19]:
df.groupby('children').count()

days_employed  dob_years  education  education_id  family_status  \
children                                                                     
-1                   44         47         47            47             47   
 0                12710      14149      14149         14149          14149   
 1                 4343       4818       4818          4818           4818   
 2                 1851       2055       2055          2055           2055   
 3                  294        330        330           330            330   
 4                   34         41         41            41             41   
 5                    8          9          9             9              9   
 20                  67         76         76            76             76   

          family_status_id  gender  income_type   debt  total_income  purpose  
children                                                                       
-1                      47      47           47     47            44       47  
 0                   14149   14149        14149  14149         12710    14149  
 1                    4818    4818         4818   4818          4343     4818  
 2                    2055    2055         2055   2055          1851     2055  
 3                     330     330          330    330           294      330  
 4                      41      41           41     41            34       41  
 5                       9       9            9      9             8        9  
 20                     76      76           76     76            67       76

The presence of values -1 and 20 in the 'children' column is considered unusual, likely resulting from input errors.

* The value -1 should be corrected to 1.
* The value 20 should be corrected to either 2 or 0.

Due to the low input quantity of these values, it is advisable to remove them for better data integrity.

In [20]:
drop_wrong_children_index = df[(df['children'] == 20) | (df['children'] == -1)].index
df.drop(drop_wrong_children_index, inplace=True)

In [21]:
df.groupby('children').count()

days_employed  dob_years  education  education_id  family_status  \
children                                                                     
0                 12710      14149      14149         14149          14149   
1                  4343       4818       4818          4818           4818   
2                  1851       2055       2055          2055           2055   
3                   294        330        330           330            330   
4                    34         41         41            41             41   
5                     8          9          9             9              9   

          family_status_id  gender  income_type   debt  total_income  purpose  
children                                                                       
0                    14149   14149        14149  14149         12710    14149  
1                     4818    4818         4818   4818          4343     4818  
2                     2055    2055         2055   2055          1851     2055  
3                      330     330          330    330           294      330  
4                       41      41           41     41            34       41  
5                        9       9            9      9             8        9

### Step 2.3. Addressing Anomalies in days_employed Column

In [22]:
data_days_employed = df.loc[:,'days_employed']

pos_count, neg_count = 0,0

for num in data_days_employed:
    if num >= 0 :
        pos_count += 1
    else:
        neg_count += 1

print('Sum of Negative Values in days_employed Column:')
print(neg_count)
print(neg_count / (neg_count + pos_count))
print()
print('Sum of Missing Values in days_employed Column:')
print(df.isnull().sum()['days_employed'])
print(df.isnull().sum()['days_employed'] / df.count()['days_employed'])

Sum of Negative Values in days_employed Column:
17971
0.8396878796374171

Sum of Missing Values in days_employed Column:
2162
0.11237006237006238


I've identified both negative and unusually large values in the 'days_employed' column. Given the significant percentage of issues in this column, a data transformation is necessary. Negative values will be converted to positive, and exceptionally large values will be excluded and replaced with the column's mean value for better accuracy.

In [23]:
df['days_employed'] = df['days_employed'].abs()

df.loc[df['days_employed'] > 36500, 'days_employed'] = pd.NaT
avg_days_employed = df['days_employed'].mean()
df['days_employed'].fillna(avg_days_employed, inplace=True)

df['years_employed'] = df['days_employed'] / 365

print(df['years_employed'])

0        23.116912
1        11.026860
2        15.406637
3        11.300677
4         6.448367
           ...    
21520    12.409087
21521     6.448367
21522     5.789991
21523     8.527347
21524     5.437007
Name: years_employed, Length: 21402, dtype: float64


In [24]:
data_days_employed = df.loc[:,'days_employed']

pos_count, neg_count = 0,0

for num in data_days_employed:
    if num >= 0 :
        pos_count += 1
    else:
        neg_count += 1

print('Sum of Negative Values in days_employed Column:')
print(neg_count)
print(neg_count / (neg_count + pos_count))
print()
print('Sum of Missing Values in days_employed Column:')
print(df.isnull().sum()['days_employed'])
print(df.isnull().sum()['days_employed'] / df.count()['days_employed'])

Sum of Negative Values in days_employed Column:
0
0.0

Sum of Missing Values in days_employed Column:
0
0.0


### Step 2.4. Dropping Abnormal Values in dob_years Column

In [25]:
dob_years_zero = df[df['dob_years'] == 0]
dob_years_zero.count().max() / df['dob_years'].count()

0.004672460517708626

The presence of zero values in the 'dob_years' column is observed. Since the percentage of these values is relatively low (0.4%), it is advisable to remove this data for better accuracy.








In [26]:
df = df[df['dob_years'] != 0].reset_index(drop=True)

In [27]:
df['dob_years'].unique()

array([42, 36, 33, 32, 53, 27, 43, 50, 35, 41, 40, 65, 54, 56, 26, 48, 24,
       21, 57, 67, 28, 63, 62, 47, 34, 68, 25, 31, 30, 20, 49, 37, 45, 61,
       64, 44, 52, 46, 23, 38, 39, 51, 59, 29, 60, 55, 58, 71, 22, 73, 66,
       69, 19, 72, 70, 74, 75])

### Step 2.5. Remove Abnormal Values in gender Column

In [28]:
df.groupby('gender')['gender'].count()

gender
F      14083
M       7218
XNA        1
Name: gender, dtype: int64

There is a single occurrence of the value 'XNA' in the 'gender' column. Due to its isolated nature, it is recommended to remove this entry for data consistency.

In [29]:
drop_wrong_gender_index = df[df['gender'] == 'XNA'].index
df.drop(drop_wrong_gender_index, inplace=True)

In [30]:
df.groupby('gender')['gender'].count()

gender
F    14083
M     7218
Name: gender, dtype: int64

### Step 2.6. Check for Duplicate Data

In [31]:
df.duplicated().sum()

71

In [32]:
df = df.drop_duplicates().reset_index(drop=True)

In [33]:
df.duplicated().sum()

0

### Step 2.7. Brief Summary of Data Transformation

In [34]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21230 entries, 0 to 21229
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21230 non-null  int64  
 1   days_employed     21230 non-null  float64
 2   dob_years         21230 non-null  int64  
 3   education         21230 non-null  object 
 4   education_id      21230 non-null  int64  
 5   family_status     21230 non-null  object 
 6   family_status_id  21230 non-null  int64  
 7   gender            21230 non-null  object 
 8   income_type       21230 non-null  object 
 9   debt              21230 non-null  int64  
 10  total_income      19149 non-null  float64
 11  purpose           21230 non-null  object 
 12  years_employed    21230 non-null  float64
dtypes: float64(3), int64(5), object(5)
memory usage: 2.1+ MB


The data transformation involves:

1. Correcting spelling errors in the 'education' column.
2. Addressing anomalies in the 'children' column.
3. Rectifying unusual values in the 'days_employed' column.
4. Managing zero values in the 'dob_years' column.
5. Removing a singular 'XNA' value in the 'gender' column.

Additionally, there is a 1.3% reduction in the dataset, contributing to enhanced data quality and reliability for subsequent analysis.

## Step 3. Working with Missing Values

### Step 3.1. Simplifying Data Processing with Family Status and Education IDs

To streamline data processing, the 'family_status' and 'education' columns have been assigned unique IDs, allowing them to be paired with a dictionary (key:value) for easier reference.


In [35]:
df[['family_status', 'family_status_id']].drop_duplicates().reset_index(drop=True)

family_status  family_status_id
0            married                 0
1  civil partnership                 1
2    widow / widower                 2
3           divorced                 3
4          unmarried                 4

In [36]:
df[['education', 'education_id']].drop_duplicates().reset_index(drop=True)

education  education_id
0    bachelor's degree             0
1  secondary education             1
2         some college             2
3    primary education             3
4      graduate degree             4

### Step 3.2. Fixing Missing Values in total_income

To address missing values in the total_income column, it is advisable to fill them with either the mean or median value of that column. The filling value can be determined by analyzing through customer age categories.

In [37]:
#Creating a Function to Categorize the Age Column
def age_group(age):
    if age < 35:
        return 'early adult'
    if 35 <= age < 45:
        return 'late adult'
    if 45 <= age <= 65:
        return 'elderly'
    else:
        return 'seniors'
    return 'Unidentified'

In [38]:
#Function Test
print(age_group(27))
print(age_group(37))
print(age_group(47))
print(age_group(67))

early adult
late adult
elderly
seniors


In [39]:
df['age_group'] = df['dob_years'].apply(age_group)

In [40]:
df.head()

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0    2353.653992         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  years_employed    age_group  
0    purchase of the house       23.116912   late adult  
1             car purchase       11.026860   late adult  
2    purchase of the house       15.406637  early adult  
3  supplementary education       11.300677  early adult  
4        to have a wedding        6.448367      elderly

Now, with the inclusion of the age category column, we'll intelligently fill in the missing values in the total_income column by utilizing the mean or median within each specific age category.







In [41]:
df_not_null = df.loc[(df['days_employed'].isna() == False) & (df['total_income'].isna() == False)]
df_not_null

children  days_employed  dob_years            education  education_id  \
0             1    8437.673028         42    bachelor's degree             0   
1             1    4024.803754         36  secondary education             1   
2             0    5623.422610         33  secondary education             1   
3             3    4124.747207         32  secondary education             1   
4             0    2353.653992         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21225         1    4529.316663         43  secondary education             1   
21226         0    2353.653992         67  secondary education             1   
21227         1    2113.346888         38  secondary education             1   
21228         3    3112.481705         38  secondary education             1   
21229         2    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21225  civil partnership                 1      F    business     0   
21226            married                 0      F     retiree     0   
21227  civil partnership                 1      M    employee     1   
21228            married                 0      M    employee     1   
21229            married                 0      F    employee     0   

       total_income                  purpose  years_employed    age_group  
0         40620.102    purchase of the house       23.116912   late adult  
1         17932.802             car purchase       11.026860   late adult  
2         23341.752    purchase of the house       15.406637  early adult  
3         42820.568  supplementary education       11.300677  early adult  
4         25378.572        to have a wedding        6.448367      elderly  
...             ...                      ...             ...          ...  
21225     35966.698     housing transactions       12.409087   late adult  
21226     24959.969        purchase of a car        6.448367      seniors  
21227     14347.610                 property        5.789991   late adult  
21228     39054.888        buying my own car        8.527347   late adult  
21229     13127.587             to buy a car        5.437007   late adult  

[19149 rows x 14 columns]

In [42]:
income_data_mean = df_not_null.groupby('income_type')['total_income'].mean()
income_data_mean

income_type
business                       32424.420789
civil servant                  27336.442546
employee                       25822.872585
entrepreneur                   79866.103000
paternity / maternity leave     8612.661000
retiree                        21950.722935
student                        15712.260000
unemployed                     21014.360500
Name: total_income, dtype: float64

In [43]:
income_data_median = df_not_null.groupby('income_type')['total_income'].median()
income_data_median

income_type
business                       27594.6410
civil servant                  24076.1150
employee                       22815.1035
entrepreneur                   79866.1030
paternity / maternity leave     8612.6610
retiree                        18959.6260
student                        15712.2600
unemployed                     21014.3605
Name: total_income, dtype: float64

In [44]:
age_group_mean = df_not_null.groupby('age_group')['total_income'].mean()
age_group_mean

age_group
early adult    26681.143124
elderly        26111.643648
late adult     28733.768065
seniors        20805.696512
Name: total_income, dtype: float64

In [45]:
age_group_median = df_not_null.groupby('age_group')['total_income'].median()
age_group_median

age_group
early adult    23406.0085
elderly        22614.0590
late adult     24818.4925
seniors        17972.4075
Name: total_income, dtype: float64


Because the data distribution is not symmetric, it is advisable to use the median as the input for missing values. It is evident that the highest values are within the late adult age group for their total income.

In [46]:
def fill_missing_total_income_value (data, agg_column, value_column):
    grouped_values = data.groupby(agg_column)[value_column].median().reset_index()
    size = len(grouped_values)
    for i in range(size):
        group = grouped_values[agg_column][i]
        value = grouped_values[value_column][i]
        data.loc[(data[agg_column]==group) & (data[value_column].isna()), value_column] = value
    return data

df = fill_missing_total_income_value(data = df, agg_column = 'age_group', value_column = 'total_income')

df['total_income'].reset_index()

index  total_income
0          0     40620.102
1          1     17932.802
2          2     23341.752
3          3     42820.568
4          4     25378.572
...      ...           ...
21225  21225     35966.698
21226  21226     24959.969
21227  21227     14347.610
21228  21228     39054.888
21229  21229     13127.587

[21230 rows x 2 columns]

In [47]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21230 entries, 0 to 21229
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21230 non-null  int64  
 1   days_employed     21230 non-null  float64
 2   dob_years         21230 non-null  int64  
 3   education         21230 non-null  object 
 4   education_id      21230 non-null  int64  
 5   family_status     21230 non-null  object 
 6   family_status_id  21230 non-null  int64  
 7   gender            21230 non-null  object 
 8   income_type       21230 non-null  object 
 9   debt              21230 non-null  int64  
 10  total_income      21230 non-null  float64
 11  purpose           21230 non-null  object 
 12  years_employed    21230 non-null  float64
 13  age_group         21230 non-null  object 
dtypes: float64(3), int64(5), object(6)
memory usage: 2.3+ MB


### Step 3.3. Fixing Values in days_employed Column

In [48]:
df_not_null.groupby('income_type')['days_employed'].median()

income_type
business                       1555.993659
civil servant                  2672.903939
employee                       1573.791064
entrepreneur                    520.848083
paternity / maternity leave    3296.759962
retiree                        2353.653992
student                         578.751554
unemployed                     2353.653992
Name: days_employed, dtype: float64

In [49]:
df_not_null.groupby('income_type')['days_employed'].mean()

income_type
business                       2119.512866
civil servant                  3392.119263
employee                       2325.740892
entrepreneur                    520.848083
paternity / maternity leave    3296.759962
retiree                        2353.653992
student                         578.751554
unemployed                     2353.653992
Name: days_employed, dtype: float64


Considering the asymmetric distribution between income_type and days_employed, it is advisable to use the median value to fill incorrect entries.

In [50]:
def fill_missing_days_employed_value (data, agg_column, value_column):
    grouped_values = data.groupby(agg_column)[value_column].median().reset_index()
    size = len(grouped_values)
    for i in range(size):
        group = grouped_values[agg_column][i]
        value = grouped_values[value_column][i]
        data.loc[(data[agg_column]==group) & (data[value_column].isna()), value_column] = value
    return data

df = fill_missing_days_employed_value(data = df, agg_column = 'income_type', value_column = 'days_employed')

df['days_employed'].reset_index()

index  days_employed
0          0    8437.673028
1          1    4024.803754
2          2    5623.422610
3          3    4124.747207
4          4    2353.653992
...      ...            ...
21225  21225    4529.316663
21226  21226    2353.653992
21227  21227    2113.346888
21228  21228    3112.481705
21229  21229    1984.507589

[21230 rows x 2 columns]

In [51]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21230 entries, 0 to 21229
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21230 non-null  int64  
 1   days_employed     21230 non-null  float64
 2   dob_years         21230 non-null  int64  
 3   education         21230 non-null  object 
 4   education_id      21230 non-null  int64  
 5   family_status     21230 non-null  object 
 6   family_status_id  21230 non-null  int64  
 7   gender            21230 non-null  object 
 8   income_type       21230 non-null  object 
 9   debt              21230 non-null  int64  
 10  total_income      21230 non-null  float64
 11  purpose           21230 non-null  object 
 12  years_employed    21230 non-null  float64
 13  age_group         21230 non-null  object 
dtypes: float64(3), int64(5), object(6)
memory usage: 2.3+ MB


## Step 4. Data Categorization

In categorizing this data, hypotheses can be used as a reference for further data processing.

1. Is there a relationship between having children and the probability of someone defaulting on a loan?
2. Is there a relationship between income level and the probability of someone defaulting on a loan?
3. How do differences in loan purposes affect the probability of someone defaulting on a loan?

Previously, some columns have already been categorized, but there are also columns that cannot be categorized.








In [52]:
df[['children', 'total_income', 'purpose']]

children  total_income                  purpose
0             1     40620.102    purchase of the house
1             1     17932.802             car purchase
2             0     23341.752    purchase of the house
3             3     42820.568  supplementary education
4             0     25378.572        to have a wedding
...         ...           ...                      ...
21225         1     35966.698     housing transactions
21226         0     24959.969        purchase of a car
21227         1     14347.610                 property
21228         3     39054.888        buying my own car
21229         2     13127.587             to buy a car

[21230 rows x 3 columns]

### Step 4.1. Children Column Categorization

In [53]:
df[['children']].value_counts()

children
0           14021
1            4792
2            2039
3             328
4              41
5               9
dtype: int64

In [54]:
def children_category (child):
    if child < 1:
        return 'no children'
    else:
        return 'has children'

In [55]:
df['children_category'] = df['children'].apply(children_category)
df

children  days_employed  dob_years            education  education_id  \
0             1    8437.673028         42    bachelor's degree             0   
1             1    4024.803754         36  secondary education             1   
2             0    5623.422610         33  secondary education             1   
3             3    4124.747207         32  secondary education             1   
4             0    2353.653992         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21225         1    4529.316663         43  secondary education             1   
21226         0    2353.653992         67  secondary education             1   
21227         1    2113.346888         38  secondary education             1   
21228         3    3112.481705         38  secondary education             1   
21229         2    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21225  civil partnership                 1      F    business     0   
21226            married                 0      F     retiree     0   
21227  civil partnership                 1      M    employee     1   
21228            married                 0      M    employee     1   
21229            married                 0      F    employee     0   

       total_income                  purpose  years_employed    age_group  \
0         40620.102    purchase of the house       23.116912   late adult   
1         17932.802             car purchase       11.026860   late adult   
2         23341.752    purchase of the house       15.406637  early adult   
3         42820.568  supplementary education       11.300677  early adult   
4         25378.572        to have a wedding        6.448367      elderly   
...             ...                      ...             ...          ...   
21225     35966.698     housing transactions       12.409087   late adult   
21226     24959.969        purchase of a car        6.448367      seniors   
21227     14347.610                 property        5.789991   late adult   
21228     39054.888        buying my own car        8.527347   late adult   
21229     13127.587             to buy a car        5.437007   late adult   

      children_category  
0          has children  
1          has children  
2           no children  
3          has children  
4           no children  
...                 ...  
21225      has children  
21226       no children  
21227      has children  
21228      has children  
21229      has children  

[21230 rows x 15 columns]

### Step 4.2. total_income Column Categorization

In [56]:
df['total_income']

0        40620.102
1        17932.802
2        23341.752
3        42820.568
4        25378.572
           ...    
21225    35966.698
21226    24959.969
21227    14347.610
21228    39054.888
21229    13127.587
Name: total_income, Length: 21230, dtype: float64

In [57]:
df['total_income'].describe()

count     21230.000000
mean      26454.234504
std       15724.945595
min        3306.762000
25%       17208.100500
50%       23319.674500
75%       31327.351000
max      362496.645000
Name: total_income, dtype: float64


Based on its distribution, it will be grouped into 4 categories, each representing 25% of the distribution.

In [58]:
def income_category (income):
    if income < 17208:
        return 'low'
    if 17208 <= income < 23319:
        return 'medium low'
    if 23319 <= income < 31327:
        return 'medium high'
    else:
        return 'high'
    return 'Unidentified'

In [59]:
df['income_category'] = df['total_income'].apply(income_category)
df

children  days_employed  dob_years            education  education_id  \
0             1    8437.673028         42    bachelor's degree             0   
1             1    4024.803754         36  secondary education             1   
2             0    5623.422610         33  secondary education             1   
3             3    4124.747207         32  secondary education             1   
4             0    2353.653992         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21225         1    4529.316663         43  secondary education             1   
21226         0    2353.653992         67  secondary education             1   
21227         1    2113.346888         38  secondary education             1   
21228         3    3112.481705         38  secondary education             1   
21229         2    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21225  civil partnership                 1      F    business     0   
21226            married                 0      F     retiree     0   
21227  civil partnership                 1      M    employee     1   
21228            married                 0      M    employee     1   
21229            married                 0      F    employee     0   

       total_income                  purpose  years_employed    age_group  \
0         40620.102    purchase of the house       23.116912   late adult   
1         17932.802             car purchase       11.026860   late adult   
2         23341.752    purchase of the house       15.406637  early adult   
3         42820.568  supplementary education       11.300677  early adult   
4         25378.572        to have a wedding        6.448367      elderly   
...             ...                      ...             ...          ...   
21225     35966.698     housing transactions       12.409087   late adult   
21226     24959.969        purchase of a car        6.448367      seniors   
21227     14347.610                 property        5.789991   late adult   
21228     39054.888        buying my own car        8.527347   late adult   
21229     13127.587             to buy a car        5.437007   late adult   

      children_category income_category  
0          has children            high  
1          has children      medium low  
2           no children     medium high  
3          has children            high  
4           no children     medium high  
...                 ...             ...  
21225      has children            high  
21226       no children     medium high  
21227      has children             low  
21228      has children            high  
21229      has children             low  

[21230 rows x 16 columns]

In [60]:
df['income_category'].value_counts()

high           5308
low            5308
medium low     5307
medium high    5307
Name: income_category, dtype: int64

## Step 5. Hypotheses Testing

### Step 5.1. Hypothesis 1: Correlation between Having Children and Probability of Loan Default

This hypothesis aims to determine whether there is a statistical correlation between individuals having children and their probability of defaulting on a loan. The analysis will explore if having children plays a significant role in predicting loan default.

In [61]:
debt_ok = df.loc[df['debt'] != 1]
debt_ng = df.loc[df['debt'] == 1]

In [62]:
def ok_percentage(column):
    return debt_ok.groupby(column)['debt'].count()*100/df.groupby(column)['debt'].count()

def ng_percentage(column):
    return debt_ng.groupby(column)['debt'].count()*100/df.groupby(column)['debt'].count()

anak = ['children_category']

for column in anak:
    print(ok_percentage(column))

children_category
has children    90.761548
no children     92.454176
Name: debt, dtype: float64


In [63]:
for column in anak:
    print(ng_percentage(column))

children_category
has children    9.238452
no children     7.545824
Name: debt, dtype: float64


There is a variation in the likelihood of loan default between a customer with children and one without. 90% of customers with children successfully repay their loans. This value is lower when compared to customers without children (92%).

By having children, the probability of loan default for a customer increases by 2%, rising from 7% to 9%.

### Step 5.2. Hypothesis 2: Correlation between Family Status and Timely Repayment

This hypothesis investigates whether there is a measurable correlation between a borrower's family status and their ability to make timely repayments. The goal is to understand if certain family statuses are associated with a higher likelihood of meeting payment deadlines.

In [64]:
status = ['family_status']
for column in status:
    print(ok_percentage(column))

family_status
civil partnership    90.685798
divorced             92.875318
married              92.442479
unmarried            90.215827
widow / widower      93.446089
Name: debt, dtype: float64


In [65]:
for column in status:
    print(ng_percentage(column))

family_status
civil partnership    9.314202
divorced             7.124682
married              7.557521
unmarried            9.784173
widow / widower      6.553911
Name: debt, dtype: float64


Among customers, the highest probability of loan default is observed in unmarried individuals, with a default rate of 9.7%. This figure shows a marginal difference compared to legally partnered customers (9.3%).

Conversely, the lowest default rate is evident among widowed customers, with an impressive loan repayment success rate of 93.4%.

### Step 5.3. Hypothesis 3: Correlation between Income Level and Timely Repayment

This hypothesis delves into the potential correlation between a borrower's income level and their ability to repay loans on time. The analysis aims to identify if individuals with different income levels demonstrate varying patterns in meeting loan repayment deadlines.

In [66]:
pendapatan = ['income_category']
for column in pendapatan:
    print(ok_percentage(column))

income_category
high           92.840995
low            91.993218
medium high    90.974185
medium low     91.709064
Name: debt, dtype: float64


In [67]:
for column in pendapatan:
    print(ng_percentage(column))

income_category
high           7.159005
low            8.006782
medium high    9.025815
medium low     8.290936
Name: debt, dtype: float64


Customers with high income exhibit the lowest default rate at 7.1%, while those with low income have the second-lowest default rate at 8.0%.

The highest default rate is observed among customers with moderate to high income, reaching 9.0%.

### Step 5.4. Hypothesis 4: Impact of Loan Purpose on Default Rate

The focus of this hypothesis is to assess how the stated purpose of a loan influences the default rate. By examining different loan purposes, the analysis seeks to uncover whether specific objectives are associated with higher or lower probabilities of loan default.






In [68]:
tujuan = ['purpose']
for column in tujuan:
    print(ok_percentage(column).sort_values())

purpose
to get a supplementary education            88.461538
purchase of a car                           88.986784
getting higher education                    89.125296
second-hand car purchase                    89.352818
profile education                           89.791183
to become educated                          90.417690
to own a car                                90.425532
cars                                        90.717300
to buy a car                                90.752688
car purchase                                90.809628
buying my own car                           90.818363
university education                        90.970655
building a property                         91.276252
going to university                         91.277890
car                                         91.581109
supplementary education                     91.611479
real estate transactions                    91.778774
getting an education                        91.799544
transactions with co

In [69]:
for column in tujuan:
    print(ng_percentage(column).sort_values())

purpose
purchase of my own house                     5.501618
housing renovation                           5.813953
property                                     6.539075
construction of own property                 6.549521
buy residential real estate                  6.844741
buy commercial real estate                   7.022901
buy real estate                              7.026144
purchase of the house for my family          7.075472
housing                                      7.244094
education                                    7.289294
housing transactions                         7.332293
purchase of the house                        7.413249
buying a second-hand car                     7.430998
to have a wedding                            7.549669
building a real estate                       7.779579
wedding ceremony                             8.025478
transactions with my real estate             8.025682
having a wedding                             8.036891
buying property for 

Based on the analysis, the highest incidence of default is linked to education-related payments (>10%). Conversely, the lowest default rates are associated with home or property purchases (5-6%).

## Step 6. Conclusion

**Conclusions Drawn from Data Processing:**

* Missing data in the provided dataset does not exhibit discernible patterns.
* Missing values are filled using either the mean or median, determined by the data distribution in each respective column.
* Duplicate entries in the data are handled by eliminating redundant records.

**Key Observations and Insights:**

* Customers with dependent children tend to have a higher probability of loan default (9.2%).
* Unmarried customers display an elevated tendency toward loan default (9.7%).
* Customers with a moderate to high income level exhibit a heightened likelihood of loan default (9.0%).
* Loans for education payments are associated with a higher default rate (>10%).
* The highest likelihood of default is observed among customers with a loan purpose for supplementary education, with a rate of 11.5%.